src/ufuncs.c

```diff
@@ -1151,6 +1156,157 @@ static void YYd_d_func(char** args, npy_intp* dimensions, npy_intp* steps, void*
 }
 static PyUFuncGenericFunction YYd_d_funcs[1] = {&YYd_d_func};
 
+
+#if GEOS_SINCE_3_9_0
+
+/* Define the geom, geom, double -> geom functions (YYd_Y) */
+static void* intersection_prec_data[1] = {GEOSIntersectionPrec_r};
+static void* difference_prec_data[1] = {GEOSDifferencePrec_r};
+static void* symmetric_difference_prec_data[1] = {GEOSSymDifferencePrec_r};
+static void* union_prec_data[1] = {GEOSUnionPrec_r};
+typedef void* FuncGEOS_YYd_Y(void* context, void* a, void* b, double c);
+static char YYd_Y_dtypes[4] = {NPY_OBJECT, NPY_OBJECT, NPY_DOUBLE, NPY_OBJECT};
+
+/* There are two inner loop functions for the YYd_Y. This is because this
+ * pattern allows for ufunc.reduce.  See YY_Y above for more information.
+ */
+static void YYd_Y_func_reduce(char** args, npy_intp* dimensions, npy_intp* steps,
+                             void* data) {
+  FuncGEOS_YYd_Y* func = (FuncGEOS_YYd_Y*)data;
+  GEOSGeometry *in1 = NULL, *in2 = NULL, *out = NULL;
+
+  // Whether to destroy a temporary intermediate value of `out`:
+  char do_destroy = 0;
+
+  GEOS_INIT_THREADS;
+
+  if (!get_geom(*(GeometryObject**)args[0], &out)) {
+    errstate = PGERR_NOT_A_GEOMETRY;
+  } else {
+    TERNARY_LOOP {
+      // Get the geometry inputs; in1 from previous iteration, in2 from array
+      in1 = out;
+      if (!get_geom(*(GeometryObject**)ip2, &in2)) {
+        errstate = PGERR_NOT_A_GEOMETRY;
+        break;
+      }
+
+      /* Either (or both) in1 and in2 could be NULL (Python: None) and in3 could be nan.
+       * Reduction operations should skip None values. We have 5 possible combinations:
+       */
+
+      double in3 = *(double*)ip3;
+
+      // 1. grid_size is nan; out will remain NULL
+      if (npy_isnan(in3)) {
+        continue;
+      }
+
+      // 2. (not NULL, not NULL); run the GEOS function
+      if ((in1 != NULL) && (in2 != NULL)) {
+        out = func(ctx, in1, in2, in3);
+
+        // Discard in1 if it was a temporary intermediate
+        if (do_destroy) {
+          GEOSGeom_destroy_r(ctx, in1);
+        }
+
+        // Mark the newly generated geometry as intermediate. Note: out will become in1.
+        do_destroy = 1;
+
+        // Break on error (we do this after discarding in1 to avoid memleaks)
+        if (out == NULL) {
+          errstate = PGERR_GEOS_EXCEPTION;
+          break;
+        }
+      }
+
+      // 3. (NULL, not NULL); When the first element of the reduction axis is None
+      else if ((in1 == NULL) && (in2 != NULL)) {
+        // Keep in2 as 'outcome' of the operation.
+        out = in2;
+        // Ensure that it will not be destroyed (it is owned by python)
+        do_destroy = 0;
+      }
+
+      // 4. (not NULL, NULL); When a None value is encountered after a not-None
+      //    Don't do `out = in1`, as that is already the case.
+      // 5. (NULL, NULL); When we have not yet encountered any not-None
+      //    Do nothing; out will remain NULL
```

Comment:
I am wondering if we could somehow reduce the duplication between `YYd_Y_func_reduce` and `YY_Y_func_reduce` ... (but my C skills are not enough to directly "see" it, except for some ugly templating, or a separate function that does the inner part and receives a flag to know if it should check in3or not or something ..) 

(now if we want this, can certainly be left for later)

Reply:
Oops - I should have deleted `YYd_Y_func_reduce` because we cannot use the `reduce` functionality with the additional `grid_size` parameter.  Gone now.